src/Stack/Config.hs

```diff
@@ -146,9 +147,23 @@ configFromConfigMonoid configStackRoot mproject configMonoid@ConfigMonoid{..} =
                 return $ progsDir </> $(mkRelDir stackProgName) </> platform
             _ -> return $ configStackRoot </> $(mkRelDir "programs") </> platform
 
-     configLocalBin <- do
-        localDir <- liftIO (getAppUserDataDirectory "local") >>= parseAbsDir
-        return $ localDir </> $(mkRelDir "bin")
+     configLocalBin <-
+         case configMonoidLocalBin of
+             Nothing -> do
+                 localDir <- liftIO (getAppUserDataDirectory "local") >>= parseAbsDir
+                 return $ localDir </> $(mkRelDir "bin")
+             Just userPath ->
+                 (liftIO $ canonicalizePath userPath >>= parseAbsDir)
+                 `catches`
+                 [Handler (\(ex :: IOException) -> error
```

Comment:
Should not be using `error`, instead throw a new [ConfigException](https://github.com/commercialhaskell/stack/blob/fc963d1f0faded90fb8ca7e5375613a0b11b67c4/src/Stack/Types/Config.hs#L515) constructor.